lib/tree.rb

```diff
@@ -16,40 +16,110 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+# add helper 
+  def helper_add(current, new_node)
+    return new_node if current.nil? 
+
+    if new_node.key <= current.key 
+      current.left = helper_add(current.left , new_node)
+    else   
+      current.right = helper_add(current.right, new_node)
+    end 
+    return current
+  end 
+
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = helper_add(@root, TreeNode.new(key,value))
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # helper method 
+  def find_helper(current, key)
+    return nil if current.nil? 
+
+    if current.key == key 
+      return current.value 
+    elsif current.key < key 
+      current = current.right 
+    else  
+      current = current.left 
+    end 
+    return find_helper(current,key)
+  end 
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current)
+    result = []
+    return [] if current.nil? 
+    result += inorder_helper(current.left)
+    result.push({:key=>current.key, :value => current.value})
+    result += inorder_helper(current.right)
+    return result
+  end 
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current)
+    result = [] 
+    return [] if current.nil? 
+    result.push({:key=>current.key, :value => current.value})
+    result += preorder_helper(current.left)
+    result += preorder_helper(current.right)
+    return result
+  end 
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # post_order_helper
+  def post_helper(current)
+    result = []
+    return [] if current.nil? 
+    result += post_helper(current.left)
+    result += post_helper(current.right)
+    result.push({:key=>current.key, :value => current.value})
+    return result
+  end 
+
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def postorder
```

Comment:
If you see my suggestion, that method has a time & space complexity of O(n), but your way due to array concatenation is O(n^2) for time complexity.